relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -23,32 +25,49 @@ import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
 import tech.relaycorp.relaynet.cogrpc.toCargoDelivery
 import tech.relaycorp.relaynet.cogrpc.toCargoDeliveryAck
 import java.io.InputStream
+import java.net.InetAddress
 import java.net.InetSocketAddress
 import java.net.URL
 import java.util.concurrent.TimeUnit
 import java.util.logging.Level
 import java.util.logging.Logger
 import kotlin.time.seconds
 
-class CogRPCClient
+open class CogRPCClient
 internal constructor(
     serverAddress: String,
-    val useTls: Boolean = true
+    val requireTls: Boolean = true
 ) {
+    private val serverUrl = URL(serverAddress)
 
-    internal val address by lazy {
-        val url = URL(serverAddress)
-        val fallbackPort = if (url.protocol == "https") 443 else 80
+    init {
+        if (requireTls && serverUrl.protocol != "https") {
+            throw CogRPCException(message = "Cannot connect to $serverAddress with TLS required")
+        }
+    }
+
+    private val address by lazy {
+        val fallbackPort = if (serverUrl.protocol == "https") 443 else 80
         InetSocketAddress(
-            url.host,
-            url.port.let { if (it != -1) it else fallbackPort }
+            serverUrl.host,
+            serverUrl.port.let { if (it != -1) it else fallbackPort }
         )
     }
 
     internal val channel by lazy {
-        NettyChannelBuilder
-            .forAddress(address)
+        val useTls = requireTls || serverUrl.protocol == "https"
+        val isHostPrivateAddress = InetAddress.getByName(serverUrl.host).isSiteLocalAddress
+        initialChannelBuilder
             .run { if (useTls) useTransportSecurity() else usePlaintext() }
+            .let { if (useTls && isHostPrivateAddress) it.sslContext(insecureTlsContext) else it }
+            .build()
+    }
+
+    internal open val initialChannelBuilder by lazy { NettyChannelBuilder.forAddress(address) }
```

Comment:
I'd really love feedback on this. I hate to have to make this field and the class open so I can inspect how the `NettyChannelBuilder` instance is used, and I thought this approach was the least horrible one -- followed by using PowerMock to mock the static method.

Normally I'd try to check the outcome instead of how a 3rd party class is used, but I don't think that's an option here:

- We don't really want to establish actual TCP connections with and without TLS, especially when some tests involve DNS hostnames on the Internet.
- We can't inspect `NettyChannelBuilder` or `ManagedChannel` instances at all, to see whether/how TLS is used.

Reply:
My suggestion is to pass a channel builder through the constructor. It could be a ChannelBuilder interface or just a `(String) -> ManagedChannel`, and it could have a default value, to avoid changing code elsewhere. It would avoid opening the class and having a public var.